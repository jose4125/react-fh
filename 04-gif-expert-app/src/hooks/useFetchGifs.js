import { useState, useEffect } from "react";
import { getGifs } from "../helpers/getGifs";

export const useFetchGifs = (category) => {
  const [images, setImages] = useState([]);
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    const loadData = async () => {
      const gifs = await getGifs(category);
      setImages(gifs);
      setLoading(false);
    };

    loadData();
  }, [category]);

  return [images, loading];
};
