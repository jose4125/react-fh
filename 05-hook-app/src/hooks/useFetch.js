import { useEffect, useRef, useState } from "react";

export const useFetch = (url) => {
  const isMounted = useRef(true);
  const [state, setState] = useState({
    data: null,
    loading: true,
  });
  const [error, setError] = useState(null);

  useEffect(() => {
    return () => {
      isMounted.current = false;
    };
  }, []);

  useEffect(
    () => {
      setState({
        data: null,
        loading: true,
      });

      const LoadData = async (url) => {
        try {
          const res = await fetch(url);
          const data = await res.json();

          if (isMounted.current) {
            setState({ data, loading: false });
          }
        } catch (error) {
          setState({ data: null, loading: false });
          setError("no se pudo cargar la data");
        }
      };
      LoadData(url);
    },
    [url]
  );

  return { state, error };
};
