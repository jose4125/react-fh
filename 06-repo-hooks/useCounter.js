import { useState } from "react";

export const useCounter = (init = 10) => {
  const [counter, setCounter] = useState(init);

  const increment = (factor = 1) => {
    setCounter(counter + factor);
  };

  const decrement = (factor = 1) => {
    setCounter(counter - factor);
  };

  const reset = () => {
    setCounter(init);
  };

  return {
    counter,
    decrement,
    increment,
    reset,
  };
};
