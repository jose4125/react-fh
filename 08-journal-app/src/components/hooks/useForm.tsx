import { useState } from 'react';

export interface FormState {
  [keyName: string]: string;
}

interface FormType {
  formState: FormState;
  handleInputChange: (
    event:
      | React.ChangeEvent<HTMLInputElement>
      | React.ChangeEvent<HTMLTextAreaElement>,
  ) => void;
  resetForm: (x: FormState) => void;
}

export const useForm = (initialState = {}): FormType => {
  const [formState, setFormState] = useState<FormState>(initialState);

  const handleInputChange = ({
    target,
  }:
    | React.ChangeEvent<HTMLInputElement>
    | React.ChangeEvent<HTMLTextAreaElement>) => {
    const key = target?.name;
    const value = target?.value;
    setFormState({ ...formState, [key]: value });
  };

  const resetForm = (newState = {}) => {
    setFormState(newState);
  };

  // const handleSubmit = e => {
  //   e.preventDefault();
  //   console.log(formState);
  // };

  return {
    formState,
    handleInputChange,
    resetForm,
    // handleSubmit,
  };
};
