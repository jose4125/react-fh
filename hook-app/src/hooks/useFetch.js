import { useEffect, useState } from "react";

export const useFetch = (url) => {
  const [state, setState] = useState({
    data: null,
    loading: true,
  });
  const [error, setError] = useState(null);

  useEffect(() => {
    setState({
      data: null,
      loading: true,
    });

    const LoadData = async (url) => {
      const res = await fetch(url);
      const data = await res.json();

      setState({ data, loading: false });
    };
    LoadData(url);
  }, [url]);

  return { state, error };
};
