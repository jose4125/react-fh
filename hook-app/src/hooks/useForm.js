import { useState } from "react";

export const useForm = (initialState = {}) => {
  const [formState, setFormState] = useState(initialState);

  const handleInputChange = ({ target }) => {
    const key = target.name;
    const value = target.value;
    setFormState({ ...formState, [key]: value });
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    console.log(formState);
  };

  return {
    formState,
    handleInputChange,
    handleSubmit,
  };
};
