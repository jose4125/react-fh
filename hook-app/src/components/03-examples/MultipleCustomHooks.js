import { useFetch } from "../../hooks/useFetch";
import { useCounter } from "../../hooks/useCounter";

export const MultipleCustomHooks = () => {
  const { counter, increment } = useCounter(1);
  const { state, error } = useFetch(
    `https://www.breakingbadapi.com/api/quotes/${counter}`
  );

  const { data, loading } = state;
  return (
    <div>
      <h1>breaking bad quotes</h1>
      <hr />
      {loading && <p>...Loading</p>}
      {data &&
        data.map(({ quote_id, quote, author }) => {
          return (
            <blockquote key={quote_id}>
              <hr />
              <p>{quote}</p>
              <footer>{author}</footer>
              <hr />
            </blockquote>
          );
        })}
      {data && <button onClick={() => increment()}>next quote</button>}
    </div>
  );
};
