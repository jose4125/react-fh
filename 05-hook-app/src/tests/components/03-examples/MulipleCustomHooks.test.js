import { shallow } from "enzyme";
import { useFetch } from "../../../hooks/useFetch";
import { MultipleCustomHooks } from "../../../components/03-examples/MultipleCustomHooks";
import { useCounter } from "../../../hooks/useCounter";

jest.mock("../../../hooks/useFetch");
jest.mock("../../../hooks/useCounter");

describe("MultipleCustomHooks tests", () => {
  beforeEach(() => {
    useCounter.mockReturnValue({
      counter: 10,
      increment: () => {},
    });
  });

  test("should render properly", () => {
    useFetch.mockReturnValue({
      state: {
        data: null,
        loading: true,
      },
      error: null,
    });

    const wrapper = shallow(<MultipleCustomHooks />);

    expect(wrapper).toMatchSnapshot();
  });

  test("Should show loading", () => {
    useFetch.mockReturnValue({
      state: {
        data: null,
        loading: true,
      },
      error: null,
    });

    const wrapper = shallow(<MultipleCustomHooks />);

    expect(wrapper.find(".loading").text()).toBe("...Loading");
    expect(wrapper.find("blockquote").exists()).toBeFalsy();
    expect(wrapper.find("button").exists()).toBeFalsy();
  });

  test("Should show data", async () => {
    useFetch.mockReturnValue({
      state: {
        data: [
          {
            quote_id: 12234,
            quote: "this is the quote",
            author: "john doe",
          },
        ],
        loading: false,
      },
      error: null,
    });

    const wrapper = shallow(<MultipleCustomHooks />);

    expect(wrapper.find(".loading").exists()).toBeFalsy();
    expect(wrapper.find("blockquote").exists()).toBeTruthy();
    expect(wrapper.find("button").exists()).toBeTruthy();
    expect(wrapper.find(".quote").exists()).toBeTruthy();
    expect(wrapper.find(".quote").text()).toBe("this is the quote");
    expect(wrapper.find("footer").text()).toBe("john doe");
    expect(wrapper.find("button").text()).toBe("next quote");
  });
});
