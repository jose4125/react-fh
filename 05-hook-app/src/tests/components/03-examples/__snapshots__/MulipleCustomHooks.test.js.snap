// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultipleCustomHooks tests should render properly 1`] = `
<div>
  <h1>
    breaking bad quotes
  </h1>
  <hr />
  <p
    className="loading"
  >
    ...Loading
  </p>
</div>
`;
